src/solver/conv_asm_implicit_gemm_bwd_v4r1_dynamic.cpp

```diff
@@ -0,0 +1,202 @@
+    int w_tilda_left = conv::igemm_dynamic::max(0, pad_w - dilation_w * (x_tilda - 1)) / stride_w;
+
+    int h_tilda_right = conv::igemm_dynamic::min(h_tilda, (pad_h + hi - 1 + stride_h - 1) / stride_h + 1);
+    int w_tilda_right = conv::igemm_dynamic::min(w_tilda, (pad_w + wi - 1 + stride_w - 1) / stride_w + 1);
+
+    int h_tilda_slice = h_tilda_right - h_tilda_left;
+    int w_tilda_slice = w_tilda_right - w_tilda_left;
+    // clang-format on
+    int gemm_m = c;
+    int gemm_n = n * h_tilda_slice * w_tilda_slice;
+    // int gemm_k; since k dimension is merged, we only check k
+
+    // TODO: this is too simple, need more kernels and more optimal logic to select kernel
+    if((gemm_m % 128 == 0) && (gemm_n % 128 == 0) && (k % 16 == 0))
+    {
+        if((y == 1) && (x == 1) && (stride_h == 1) && (stride_w == 1) && (dilation_h == 1) &&
+           (dilation_w == 1) && (pad_h == 0) && (pad_w == 0) && (n % 128 == 0))
+        {
+            grid_size   = (gemm_m >> 7) * (gemm_n >> 7);
+            block_size  = 256;
+            kernel_name = "igemm_bwd_gtc_bt128x128x16_tt8x8_gm2x4x4_gn2x4x4_ta1x1x1x2x4_"
+                          "16x1x1x16x1_tb1x1x1x2x4x1x1_16x1x1x16x1x1x1";
+            return true;
+        }
+        else
+        {
+            grid_size   = (gemm_m >> 7) * (gemm_n >> 7);
+            block_size  = 256;
+            kernel_name = "igemm_bwd_gtc";
+            return true;
+        }
+    }
+    else
+    {
+        if((y == 1) && (x == 1) && (stride_h == 1) && (stride_w == 1) && (dilation_h == 1) &&
+           (dilation_w == 1) && (pad_h == 0) && (pad_w == 0))
+        {
+            if((gemm_m % 128 == 0) && (gemm_n % 128 == 0) && (k % 8 == 0) && ((ho * wo) % 16 == 0))
+            {
+                grid_size   = (gemm_m >> 7) * (gemm_n >> 7);
+                block_size  = 256;
+                kernel_name = "igemm_bwd_gtc_bt128x128x8_tt8x8_gm2x4x4_gn2x4x4_ta1x1x1x1x4_"
+                              "8x1x1x32x1_tb1x1x1x1x4x1x1_8x1x1x2x1x1x16";
+                return true;
+            }
+            else if((gemm_m % 64 == 0) && (gemm_n % 64 == 0) && (k % 8 == 0) && (n % 64 == 0))
+            {
+                grid_size   = (gemm_m >> 6) * (gemm_n >> 6);
+                block_size  = 64;
+                kernel_name = "igemm_bwd_gtc_bt64x64x8_tt8x8_gm2x4x2_gn2x4x2_ta1x2x1x1x4_"
+                              "4x1x1x16x1_tb1x2x1x1x4x1x1_4x1x1x16x1x1x1";
+                return true;
+            }
+        }
+        else
+        {
+            // TODO: add here!
+        }
+    }
+    return false;
+}
+
+bool ConvAsmImplicitGemmV4R1DynamicBwd::IsApplicable(const ConvolutionContext& ctx) const
+{
+    const auto device_name = ctx.GetStream().GetDeviceName();
+    if(!(StartsWith(device_name, "gfx900") || StartsWith(device_name, "gfx906")))
+        return false;
+
+    if(!ctx.direction.IsBackwardData())
+        return false;
+
+    if(!ctx.Is2d())
+        return false;
+
+    if(!ctx.IsFp32())
```

Comment:
What's the plan to support fp16?

Reply:
There do have plan for gfx908+fp16, which I'm going to implement next. and according to priority currently no plan for vega20+fp16.